walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
 
 public final class FindMeetingQuery {
+
+  /**
+  * Determines which attendees to include and returns time ranges that accomadate those attendees
+  */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+
+    // Return all times if there are no attendees
+    if (request.getAttendees().isEmpty() && request.getOptionalAttendees().isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } 
+
+    ArrayList<String> mandatoryAttendees = new ArrayList<String>(request.getAttendees());
+    ArrayList<String> allAttendees = new ArrayList<String>(mandatoryAttendees);
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    ArrayList<TimeRange> timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(allAttendees, events, duration));
+
+    // If no time ranges are available, exclude optional attendees and try again
+    if (!timeRanges.isEmpty()) {
+      return timeRanges;
+    }
+
+    // If the request has no attendees after optional attendees are excluded, return an empty collection.
+    if (!mandatoryAttendees.isEmpty()) {
+      timeRanges = new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));
```

Comment:
nit: you can just do 
return new ArrayList<TimeRange>(findAvailableTimeRanges(mandatoryAttendees, events, duration));

Reply:
done.